src/client/Microsoft.Identity.Client/Internal/ClientCredentialWrapper.cs

```diff
@@ -150,43 +72,132 @@ private void ValidateCredentialParameters(ApplicationConfiguration config)
+            ApplicationConfiguration applicationConfiguration = new ApplicationConfiguration();
+            applicationConfiguration.ClientCredentialCertificate = certificate;
+            applicationConfiguration.ConfidentialClientCredentialCount = 1;
+            applicationConfiguration.ClaimsToSign = claimsToSign;
+
+            return new ClientCredentialWrapper(applicationConfiguration);
         }
 
+        public static ClientCredentialWrapper CreateWithSecret(string secret)
+        {
+            ApplicationConfiguration applicationConfiguration = new ApplicationConfiguration();
+            applicationConfiguration.ClientSecret = secret;
+            applicationConfiguration.ConfidentialClientCredentialCount = 1;
+
+            return new ClientCredentialWrapper(applicationConfiguration);
+        }      
+
+        #endregion TestBuilders       
+   
         public static int MinKeySizeInBits { get; } = 2048;
-        internal string Thumbprint { get { return Base64UrlHelpers.Encode(Certificate.GetCertHash()); } }
-        internal X509Certificate2 Certificate { get; private set; }
+        internal string Thumbprint { get; }
+        internal X509Certificate2 Certificate { get; }
         // The cached assertion created from the JWT signing operation
-        internal string CachedAssertion { get; set; }
-        internal long ValidTo { get; set; }
-        internal bool ContainsX5C { get; set; }
-        internal string Audience { get; set; }
-        internal string Secret { get; private set; }
+        internal string Secret { get; }
         // The signed assertion passed in by the user
-        internal string SignedAssertion { get; private set; }
-        internal Func<string> SignedAssertionDelegate { get; private set; }
-        internal bool AppendDefaultClaims { get; private set; }
-        internal ConfidentialClientAuthenticationType AuthenticationType { get; private set; }
-        internal IDictionary<string, string> ClaimsToSign { get; private set; }
-    }
+        internal string SignedAssertion { get; }
+        internal Func<string> SignedAssertionDelegate { get; }
+        internal bool AppendDefaultClaims { get;  }
+        internal ConfidentialClientAuthenticationType AuthenticationType { get;  }
+        internal IDictionary<string, string> ClaimsToSign { get; }
+
+        public void AddConfidentialClientParameters(
+            OAuth2Client oAuth2Client,
+            ICoreLogger logger,
+            ICryptographyManager cryptographyManager,
+            string clientId,
+            Authority authority,
+            bool sendX5C)
+        {
+            using (logger.LogMethodDuration())
+            {
+                switch (AuthenticationType)
+                {
+                    case ConfidentialClientAuthenticationType.ClientCertificate:
+                        string tokenEndpoint = authority.GetTokenEndpoint();
+
+                        var jwtToken2 = new JsonWebToken(
```

Comment:
can this return a null value or throw?

Reply:
I don't think so, it's just a data object, it copies those values in the ctor and keeps a reference to the crypto manager so that later on it can sign the jwt.